File: web/lib/wordCount.ts
// 中英文字数统计工具
//
// 功能概述:
// 提供中英混合文本的字数统计能力。中文字符按字计数,
// 连续英文字母序列按单词计数。
// 与桌面版 wordCounter 逻辑保持一致, 但独立实现以避免引用桌面版。
//
// 模块职责:
// 1. countWords: 统计纯文本字数
// 2. countWordsFromProseMirror: 从 ProseMirror JSON 文档统计字数

import type { ProseMirrorNode } from "./types";

/**
 * 统计文本中的字数 (中文字符 + 英文单词)
 * 中文字符范围: U+4E00~U+9FFF, U+3400~U+4DBF, U+F900~U+FAFF (兼容汉字)
 * 英文单词: 连续英文字母序列计为一个单词
 *
 * @param text 待统计文本
 * @returns 字数
 */
export function countWords(text: string): number {
  if (!text) return 0;

  let count = 0;
  let inWord = false;

  for (const ch of text) {
    if (
      ("\u4E00" <= ch && ch <= "\u9FFF") ||
      ("\u3400" <= ch && ch <= "\u4DBF") ||
      ("\uF900" <= ch && ch <= "\uFAFF")
    ) {
      // 中文字符: 每字计 1
      count += 1;
      inWord = false;
    } else if (/[a-zA-Z]/.test(ch)) {
      // 英文字母: 连续字母序列计为 1 个单词
      if (!inWord) {
        count += 1;
        inWord = true;
      }
    } else {
      // 非中非英字符: 重置英文单词状态
      inWord = false;
    }
  }

  return count;
}

/**
 * 从 ProseMirror JSON 文档提取所有文本并统计字数
 *
 * @param doc ProseMirror 文档节点
 * @returns 字数
 */
export function countWordsFromProseMirror(doc: ProseMirrorNode): number {
  const text = extractTextFromNode(doc);
  return countWords(text);
}

/**
 * 递归提取 ProseMirror 节点中的所有纯文本
 *
 * @param node ProseMirror 节点
 * @returns 拼接后的纯文本
 */
function extractTextFromNode(node: ProseMirrorNode): string {
  // 文本节点: 直接返回 text 字段
  if (node.type === "text" && node.text) {
    return node.text;
  }
  // 容器节点: 递归拼接子节点文本
  if (node.content && node.content.length > 0) {
    return node.content.map((child) => extractTextFromNode(child)).join("");
  }
  return "";
}

File: src-tauri/src/word_count.rs
// 共享字数统计模块
//
// 功能概述：
// 提供统一的中英文字数统计函数，供 commands 与 snapshot_commands 复用，
// 消除原先在多个模块中重复实现的 count_chinese_and_words / count_words_simple。
//
// 模块职责：
// 1. 统一中文汉字 + 英文单词的计数规则
// 2. 供其他 Rust 模块通过 use crate::word_count::count_words 调用
//
// 计数规则：
// - 中文字符（基本汉字区 U+4E00..U+9FFF、扩展A区 U+3400..U+4DBF、兼容汉字区 U+F900..U+FAFF）按 1 字计算
// - 英文连续字母序列作为一个单词按 1 字计算
// - 数字、标点、空白等不计入

/// 统计中文字符与英文单词数
/// 输入: text 文本内容
/// 输出: u64 字数
/// 流程: 遍历字符，中文字符计数，英文连续字母作为一个单词
pub fn count_words(text: &str) -> u64 {
    let mut count: u64 = 0;
    let mut in_word = false;
    for ch in text.chars() {
        // 中文字符范围(基本汉字 + 扩展A + 兼容汉字)
        if ('\u{4E00}'..='\u{9FFF}').contains(&ch)
            || ('\u{3400}'..='\u{4DBF}').contains(&ch)
            || ('\u{F900}'..='\u{FAFF}').contains(&ch)
        {
            count += 1;
            in_word = false;
        } else if ch.is_alphabetic() {
            if !in_word {
                count += 1;
                in_word = true;
            }
        } else {
            in_word = false;
        }
    }
    count
}
